### PR TITLE
messaging: support vt_message_cols to limit cols

### PR DESCRIPTION
**Pull request by derekperkins**
## Description
this enables users to set an allow-list of columns that will be loaded into memory and streamed to subscribers when they call `stream * from tbl`

## Related Issue(s)
#9666 

## Checklist
- [x] Should this PR be backported? No
- [x] Tests were added
- [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by mattlord** [comment]
@derekperkins I apologize for the delay, I will be reviewing this today.

**Comment by derekperkins** [comment]
Great, thanks. I'll clean it up and get it passing tests again. I've been using this as our custom build branch trying to solve #9936. We've had this running in prod for 6 weeks or so and it has been working great.

File: go/test/endtoend/messaging/message_test.go
Comment (mattlord) [inline go/test/endtoend/messaging/message_test.go:247]: Don't feel strongly, but I think `vt_message_cols` would be better since this applies to the messages at a higher level, and at the lower level applies to both the poller and the results stream.

And beyond that ~ `createLimitedStreamingColsMessage` or ~ `TestSpecificStreamingColsMessage` would be slightly better IMO. Also minor things that I don't feel strongly about.

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/load_table.go:131]: This is where `allowed` feels misleading as you're saying that they are required. If they were allowed then I would expect that them not being present would be fine.

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/load_table.go:140]: But I guess we don't need the other requiredCols here? 
```
		"priority",
 		"time_next",
 		"epoch",
 		"time_acked",
```

I guess those are required in the table structure but not the messages... that would make sense. This could be a reason that `vt_message_fields` is the best name...

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/load_table.go:180]: personal preference but after so many years my brain sees the semi-colon in SQL as a hard stmt/context delimiter. I can see that comma has its own problems in this context, but I think I would prefer pipe as it's not uncommon to use that as a column separator in output.

Another thing I don't feel super strongly about and leave up to you to make the final call on.

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/load_table.go:175]: I feel like this should be called `parseMessageFields`?  Minor thing, I just didn't realize what it was for at first. We could also address that with a function comment.

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/load_table.go:202]: This is case and white space sensitive, which is fine, probably just worth documenting as most identifier comparisons in SQL are not (or dependent on the collation used). We could potentially use the new evalengine here to compare the identifiers in the same way MySQL would.

**Review by mattlord — APPROVED** [review]
Looks good. Thank you for working on this! This will make it feasible to have much larger message cache sizes.

I had some questions/comments/nits mostly related to naming but you can make the final call on the naming aspects. Naming things is hard, and as long as we update the docs and explain the field well that's what matters most. Speaking of that, I assume you'll have a docs PR ready to review as well? Let me know. That has gotten much easier now that we have versioned docs and I can offer help if needed.

**Review by rohit-nayak-ps — COMMENTED** [review]
aside from minor nits, functionality lgtm

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (mattlord) [reply go/vt/vttablet/tabletserver/schema/load_table.go:202]: To explain what I mean here, it would be something like this:
```
func getAllowedMessageFields(tableFields []*querypb.Field, streamCols []string) []*querypb.Field {
	fields := make([]*querypb.Field, 0, len(streamCols))
	// Load user-defined columns. Any "unrecognized" column is user-defined.
	for _, col := range streamCols {
		for _, field := range tableFields {
			if res, _ := evalengine.NullsafeCompare(sqltypes.NewVarChar(field.Name), sqltypes.NewVarChar(col), collations.Default()); res == 0 {
				fields = append(fields, field)
				break
			}
		}
	}
	return fields
}
```

Comment (mattlord) [reply go/vt/vttablet/tabletserver/schema/load_table.go:202]: And if we wanted to deal with the user potentially having whitespace in the comment value:
```
if res, _ := evalengine.NullsafeCompare(sqltypes.NewVarChar(strings.TrimSpace(field.Name)), sqltypes.NewVarChar(strings.TrimSpace(col)), collations.Default()); res == 0 {
				fields = append(fields, field)
				break
			}
```

Comment (derekperkins) [reply go/vt/vttablet/tabletserver/schema/load_table.go:202]: Perfect, I'll go ahead and use that. I may clean up the other functions here in a secondary commit too

Comment (mattlord) [reply go/vt/vttablet/tabletserver/schema/load_table.go:202]: Cool! To be clear, this is what I was thinking for the final function (ignoring the function name which is a separate issue):
```
func getAllowedMessageFields(tableFields []*querypb.Field, streamCols []string) []*querypb.Field {
	fields := make([]*querypb.Field, 0, len(streamCols))
	// Load user-defined columns. Any "unrecognized" column is user-defined.
	for _, col := range streamCols {
		for _, field := range tableFields {
			if res, _ := evalengine.NullsafeCompare(sqltypes.NewVarChar(field.Name),
				sqltypes.NewVarChar(strings.TrimSpace(col)),
				collations.Default()); res == 0 {
				fields = append(fields, field)
				break
			}
		}
	}
	return fields
}
```

This way the comparison is case insensitive as the default collation with MySQL 5.7 is `utf8mb4_general_ci` and with 8.0 is `utf8mb4_0900_ai_ci` and we're stripping any leading or trailing whitespace from the identifiers in the comment directive as I can see that happening (in a list you may expect whitespace not to matter).

File: go/test/endtoend/messaging/message_test.go
Comment (derekperkins) [reply go/test/endtoend/messaging/message_test.go:247]: done

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (derekperkins) [reply go/vt/vttablet/tabletserver/schema/load_table.go:131]: renamed to `specifiedCols`

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (derekperkins) [reply go/vt/vttablet/tabletserver/schema/load_table.go:175]: done

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (derekperkins) [reply go/vt/vttablet/tabletserver/schema/load_table.go:178]: Good call. I was just mirroring the other functions, but that was overkill. Fixed

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (derekperkins) [reply go/vt/vttablet/tabletserver/schema/load_table.go:180]: done

File: go/vt/vttablet/tabletserver/schema/load_table.go
Comment (derekperkins) [reply go/vt/vttablet/tabletserver/schema/load_table.go:202]: done

File: go/vt/vttablet/tabletserver/schema/load_table_test.go
Comment (derekperkins) [reply go/vt/vttablet/tabletserver/schema/load_table_test.go:172]: there were already tests above without ~vt_stream_cols~`vt_message_cols` specified that I left unchanged to handle that case

**Comment by derekperkins** [comment]
Review comments are all addressed now in the last commit in the last 2 commits. This is rebased on top of #10132. Once that is fixed and merged, I'll rebase again to get the tests passing

**Comment by mattlord** [comment]
> Review comments are all addressed now in the last commit in the last 2 commits. This is rebased on top of #10132. Once that is fixed and merged, I'll rebase again to get the tests passing

@derekperkins Should be good now, test wise. I am still working on the locking related stuff in the other PR but the tests should be done.

**Comment by derekperkins** [comment]
@mattlord I think this should be good to go. I ran into some rebasing issues when switching from being rebased on #10132 to main, not realizing you had squashed. In the end, it was easiest for me to just squash mine too, vs leaving the separate review commit. I addressed all the review items, via name changes and comments. May be worth a quick once over again, but no major functionality changes.

File: go/test/endtoend/messaging/message_test.go
Comment (mattlord) [inline go/test/endtoend/messaging/message_test.go:274]: Any reasons not to use the new structure we have in the docs? Not a big deal, but I'd like to avoid drift between what we test and what we tell users to use if possible from here on out.

Comment (derekperkins) [reply go/test/endtoend/messaging/message_test.go:274]: Just didn't think to change it. I'll update the tests

**Review by mattlord — APPROVED** [review]
Looks great! So much easier to read for me. Thank you for taking the time to make those changes, add comments, etc. It's not only easier to read, but more resilient and easier to use as well. ❤️ 


I had one minor question about the table structure used — hoping we can base it on what we now have in the docs — but that's it.

Thanks again!

**Comment by mattlord** [comment]
Oh, one last thing... will you be opening a website PR as well so that we add details on this new configuration parameter to the v14 docs here: https://github.com/vitessio/website/blob/prod/content/en/docs/14.0/reference/features/messaging.md and here: https://github.com/vitessio/website/blob/prod/content/zh/docs/14.0/reference/messaging.md ?

I can help with that (creation or review), just let me know.

**Comment by mattlord** [comment]
Friendly ping on this 🙂

File: go/test/endtoend/messaging/message_test.go
Comment (derekperkins) [reply go/test/endtoend/messaging/message_test.go:276]: done

**Comment by derekperkins** [comment]
Working on the docs PR now

**Pull request merged**
